File: LeetCode-C++/Remove Nth Node From End of List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *	 int val;
 *	 ListNode *next;
 *	 ListNode(int x) : val(x), next(NULL) {}
 * };
 */

#define F(i, a, b) for( int i = a; i <= b; i++ )

class Solution {
public:
	//typedef
	typedef ListNode* iterator;

	//methods
	ListNode *removeNthFromEnd(ListNode *head, int n) {
		if( head == NULL || n < 1 ){
			return head;
		}

		iterator phead = new ListNode(0);
		phead->next = head;

		if( n == 1 ){
			iterator sl = phead;
			iterator fst = sl->next;
			while( fst->next != NULL ){
				sl = fst;
				fst = fst->next;
			}
			sl->next = NULL;

			return phead->next;
		}else{
			iterator sl = phead;
			iterator fst = sl->next;
			F( i, 1, n-1 ){
				fst = fst->next;
			}
			while( fst->next != NULL ){
				sl = sl->next;
				fst = fst->next;
			}
			fst = sl->next;
			sl->next = fst->next; 

			return phead->next;
		}
	}
};

File: LeetCode-Java/Remove Nth Node From End of List.java




public class Solution {
	public ListNode removeNthFromEnd(ListNode head, int n) {
		if( head == null || n < 1 ){
			return head;
		}

		ListNode phead = new ListNode(0);
		phead.next = head; 
		if( n == 1 ){
			ListNode slo = phead; 
			ListNode fst = head; 
			while( fst.next != null ){
				slo = fst;
				fst = fst.next;
			}
			slo.next = null;

			return phead.next;
		}else{
			if( head.next == null ){
				return head;
			}
			ListNode pslo = phead; 
			ListNode slo = head; 
			ListNode fst = head.next;

			//move right for n-2 steps
			for( int i = 1; i <= n-2; i++ ){
				if( fst.next == null ){
					return head; 
				}else{ 
					fst = fst.next;
				}
			}

			while( fst.next != null ){
				pslo = slo;
				slo = slo.next;
				fst = fst.next; 
			}
			pslo.next = slo.next; 
		}

		return phead.next; 
	}
}



